### PR TITLE
Platform/windows arm64

### PR DESCRIPTION
**Pull request by jem-patel**
- Should this pull request have release notes?
  - [x] Yes
  - [ ] No
- Do these changes need to be back ported?
  - [ ] Yes
  - [x] No
- Do these changes need to be upstreamed to [mono/mono](https://github.com/mono/mono) or [dotnet/runtime](https://github.com/dotnet/runtime) repositories?
  - [ ] Yes
  - [ ] No

Reviewers: please consider these questions as well! :heart:

https://user-images.githubusercontent.com/56306816/206303985-940d948c-60a6-4989-80c7-02bb41618f41.mp4

Testing for Self & Tom: 
Build: https://unity-ci.cds.internal.unity3d.com/job/19243605 
Unity : https://github.cds.internal.unity3d.com/unity/unity/pull/19774
Katana: https://katana.ds.unity3d.com/projects/Unity/builders?unity_branch=platform%2Fwindows%2Farm64%2Fmono

**Release notes**

Internal @jem-patel:
Mono: Add support for Windows ARM64

https://jira.unity3d.com/browse/POI-600

### CODE REVIEW

File: mono/metadata/coree.c [hunk 1]
```diff
@@ -630,7 +630,7 @@ STDAPI MonoFixupCorEE(HMODULE ModuleHandle)
 #if defined(_M_IA64)
 			ProcRva = (DWORD)((DWORD_PTR)PLabel - (DWORD_PTR)DosHeader);
 			*(PLabel)++ = *ExportFixup->ProcAddress.PLabel;
-#elif defined(_M_AMD64)
+#elif defined(_M_AMD64) || defined(_M_ARM64)
```
Comment (joncham) [inline mono/metadata/coree.c]: I would suggest not defining `HAVE_API_SUPPORT_WIN32_COREE` for ARM64.

See where it's defined here: https://github.com/Unity-Technologies/mono/blob/ffd005d73712c4aa5fd2d5f36105c79855754e40/mono/utils/w32subset.h#L81

File: mono/mini/mini-ops.h [hunk 2]
```diff
@@ -1382,7 +1382,7 @@ MINI_OP(OP_PPC_FLOOR,              "ppc_floor", FREG, FREG, NONE)
 MINI_OP(OP_PPC_TRUNC,              "ppc_trunc", FREG, FREG, NONE)
 #endif
 
-#if defined(TARGET_ARM) || defined(TARGET_ARM64)
+#if defined(TARGET_ARM) || defined(TARGET_ARM64) || defined(TARGET_AMD64)
```
Comment (joncham) [inline mono/mini/mini-ops.h]: I don't think `TARGET_AMD64` should enable this code path.

File: mono/mini/mini-ops.h [hunk 3]
```diff
@@ -1499,7 +1499,7 @@ MINI_OP(OP_MIPS_COND_EXC_INC, "mips_cond_exc_inc", NONE, IREG, IREG)
 
 #endif
 
-#if defined(TARGET_ARM64)
+#if defined(TARGET_ARM64) || defined(TARGET_AMD64)
```
Comment (joncham) [inline mono/mini/mini-ops.h]: Same as above, these are arm specific instructions.

Comment (jem-patel) [reply mono/mini/mini-ops.h]: (same comment as above) so the reason for this is because i have to build the genmdesc project for windows x64 to build arm64 files which means the TARGET_ARM64 is not set 

so the build fails as those are not included for the files created

File: mono/mini/mini-ops.h (hunk 2 above)
Comment (jem-patel) [reply mono/mini/mini-ops.h]: so the reason for this is because i have to build the genmdesc project for windows x64 to build arm64 files which means the TARGET_ARM64 is not set

so the build fails as those are not included for the files created

File: external/buildscripts/build_win_no_cygwin.pl [hunk 4]
```diff
@@ -153,10 +164,33 @@
...
 
-	my $embedDirArchDestination = $arch32 ? "$embedDirRoot\\win32" : "$embedDirRoot\\win64";
-	my $distDirArchBin = $arch32 ? "$distdir\\bin" : "$distdir\\bin-x64";
-	my $versionsOutputFile = $arch32 ? "$buildsroot\\versions-win32.txt" : "$buildsroot\\versions-win64.txt";
-
+	my $embedDirArchDestination = "";
+   	my $distDirArchBin = "";
+   	my $versionsOutputFile = "";
+	
+	print(">>> TARGET ARCH IS $targetArch\n");
```
Comment (jem-patel) [inline external/buildscripts/build_win_no_cygwin.pl]: will remove in next iteration

File: mono/mini/mini-ops.h (hunk 2 above)
Comment (joncham) [reply mono/mini/mini-ops.h]: I think this maybe the distinction between HOST_ARM64 (platform the executable will run on) vs TARGET_ARM64 (target for machine code generated by the executable).

Mono wasn't always precise with this in past, especially on Windows, but this may be the distinction needed. I'll take a closer look tomorrow.

File: external/buildscripts/build.pl [hunk 5]
```diff
@@ -1536,9 +1535,9 @@
 	}
 	else
 	{
-		$embedDirArchDestination = $arch32 ? "$embedDirRoot/win32" : "$embedDirRoot/win64";
-		$distDirArchBin = $arch32 ? "$distdir/bin" : "$distdir/bin-x64";
-		$versionsOutputFile = $arch32 ? "$buildsroot/versions-win32.txt" : "$buildsroot/versions-win64.txt";
+		$embedDirArchDestination = "$embedDirRoot/$targetArch";
```
Comment (jem-patel) [inline external/buildscripts/build.pl]: fixed in next interation

File: mono/mini/mini-arm64.c [hunk 6]
```diff
@@ -2040,7 +2045,7 @@ mono_arch_flush_icache (guint8 *code, gint size)
 #ifndef MONO_CROSS_COMPILE
 #if __APPLE__
 	sys_icache_invalidate (code, size);
-#else
+#elif !HOST_WIN32
```
Comment (TautvydasZilys) [inline mono/mini/mini-arm64.c]: You are missing Win32 path here. Use [FlushInstructionCache](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-flushinstructioncache).

File: external/buildscripts/build_win_no_cygwin.pl [hunk 7]
```diff
@@ -121,22 +125,29 @@
...
+		copy("$monoroot/msvc/build/sgen/$archNameForBuild/bin/$configDirName/mono-2.0-sgen.dll", "$monoprefix/bin/.") or die ("failed copying mono-2.0-sgen.dll\n");
+		copy("$monoroot/msvc/build/sgen/$archNameForBuild/bin/$configDirName/mono-2.0-sgen.pdb", "$monoprefix/bin/.") or die ("failed copying mono-2.0-sgen.pdb\n");
+		
+		# sgen as default exe
+		copy("$monoroot/msvc/build/sgen/$archNameForBuild/bin/$configDirName/mono-sgen.exe", "$monoprefix/bin/mono.exe") or die ("failed copying mono-sgen.exe to mono.exe\n");
+	}
+	else
+	{
+		# bdwgc as default exe
+		copy("$monoroot/msvc/build/boehm/$archNameForBuild/bin/$configDirName/mono-bdwgc.exe", "$monoprefix/bin/mono.exe") or die ("failed copying mono-bdwgc.exe to mono.exe\n");
```
Comment (TautvydasZilys) [inline external/buildscripts/build_win_no_cygwin.pl]: Do we even need the .exe at all for ARM64?

File: external/buildscripts/build_win_no_cygwin.pl [hunk 8]
```diff
@@ -199,22 +231,28 @@
 
 	copy("$monoprefix/bin/mono-2.0-bdwgc.dll", "$embedDirArchDestination/.") or die ("failed copying mono-2.0-bdwgc.dll\n");
 	copy("$monoprefix/bin/mono-2.0-bdwgc.pdb", "$embedDirArchDestination/.") or die ("failed copying mono-2.0-bdwgc.pdb\n");
```
Comment (TautvydasZilys) [inline external/buildscripts/build_win_no_cygwin.pl:233]: This is also not needed for ARM64, I think.

File: mono/mini/mini-arm64.h [hunk 9]
```diff
@@ -15,7 +15,89 @@
...
+	guint64 eax;
+	guint64 ebx;
+	guint64 ecx;
+	guint64 edx;
+	guint64 ebp;
+	guint64 esp;
+	guint64 esi;
+	guint64 edi;
+	guint64 eip;
+};
```
Comment (TautvydasZilys) [inline mono/mini/mini-arm64.h]: This looks very wrong. These are x86 registers. What are they doing here?

File: mono/mini/mini-arm64.h [hunk 10]
```diff
@@ -15,7 +15,89 @@
...
+	DWORD MaximumEntryCount,
+	ULONG_PTR RangeBase,
+	ULONG_PTR RangeEnd);
+
+typedef VOID(NTAPI* RtlGrowFunctionTablePtr)(
+	PVOID DynamicTable,
+	DWORD NewEntryCount);
+
+typedef VOID(NTAPI* RtlDeleteGrowableFunctionTablePtr)(
+	PVOID DynamicTable);
```
Comment (TautvydasZilys) [inline mono/mini/mini-arm64.h]: These are currently only called from `exceptions-amd64.c` which will not be called on ARM64. I think those calls need to be moved to `mini-windows.c` so it can be used on both x64 and ARM64.

File: mono/mini/mini-windows.c [hunk 11]
```diff
@@ -187,7 +187,7 @@ install_custom_handler (const char *handlers, size_t *handler_arg_len)
 void
 mono_runtime_install_handlers (void)
 {
-#ifndef MONO_CROSS_COMPILE
+#if !defined(MONO_CROSS_COMPILE) && !defined(TARGET_ARM64)
 	win32_seh_init();
 	win32_seh_set_handler(SIGFPE, mono_sigfpe_signal_handler);
 	win32_seh_set_handler(SIGILL, mono_crashing_signal_handler);
```
Comment (TautvydasZilys) [inline mono/mini/mini-windows.c]: @joncham correct me if I'm wrong, but if these are not installed, won't the runtime fail to deal with null reference exceptions and such?

I think x64 implementation can be reused here, by moving stuff from `exceptions-amd64.c` to this file and guarding it with `#if _WIN64` or something.

File: mono/utils/winarm64.asm [hunk 12]
```diff
@@ -0,0 +1,48 @@
+    AREA    |.text|,CODE
+
+    EXPORT  |mono_context_get_current|
+    
+|mono_context_get_current|  PROC
+		mov x16, x0
+		stp     x0, x1, [x16], #16
```
Comment (TautvydasZilys) [inline mono/utils/winarm64.asm]: nit: tabs vs spaces.

File: external/buildscripts/build_win_no_cygwin.pl [hunk 13]
```diff
@@ -199,22 +231,28 @@
...
+		copy("$monoprefix/bin/mono-2.0-sgen.pdb", "$embedDirArchDestination/.") or die ("failed copying mono-2.0-sgen.pdb\n");
+	}
+	
 	copy("$monoprefix/bin/MonoPosixHelper.dll", "$embedDirArchDestination/.") or die ("failed copying MonoPosixHelper.dll\n");
 	copy("$monoprefix/bin/MonoPosixHelper.pdb", "$embedDirArchDestination/.") or die ("failed copying MonoPosixHelper.pdb\n");
 
 	# monodistribution directory setup
 	print(">>> Creating monodistribution directory\n");
 	copy("$monoprefix/bin/mono-2.0-bdwgc.dll", "$distDirArchBin/.") or die ("failed copying mono-2.0-bdwgc.dll\n");
 	copy("$monoprefix/bin/mono-2.0-bdwgc.pdb", "$distDirArchBin/.") or die ("failed copying mono-2.0-bdwgc.pdb\n");
```
Comment (TautvydasZilys) [inline external/buildscripts/build_win_no_cygwin.pl]: These are also not needed for ARM64, I think.

File: external/buildscripts/build_win_no_cygwin.pl [hunk 14]
```diff
@@ -199,22 +231,28 @@
...
+	
+	if($targetArch ne 'ARM64')
+    {
+		copy("$monoprefix/bin/mono-2.0-sgen.dll", "$distDirArchBin/.") or die ("failed copying mono-2.0-sgen.dll\n");
+		copy("$monoprefix/bin/mono-2.0-sgen.pdb", "$distDirArchBin/.") or die ("failed copying mono-2.0-sgen.pdb\n");
+	
+		copy("$monoprefix/bin/mono-sgen.exe", "$distDirArchBin/.") or die ("failed copying mono-sgen.exe\n");
+	}
 	copy("$monoprefix/bin/mono-bdwgc.exe", "$distDirArchBin/.") or die ("failed copying mono-bdwgc.exe\n");
 	copy("$monoprefix/bin/mono.exe", "$distDirArchBin/.") or die ("failed copying mono.exe\n");
```
Comment (TautvydasZilys) [inline external/buildscripts/build_win_no_cygwin.pl]: And these.

File: mono/mini/mini-arm64.c [hunk 15]
```diff
@@ -2064,6 +2069,8 @@ mono_arch_flush_icache (guint8 *code, gint size)
 
 	asm volatile ("dsb ish" : : : "memory");
 	asm volatile ("isb" : : : "memory");
+#elif HOST_WIN32
```
Comment (joncham) [inline mono/mini/mini-arm64.c]: You could invert these checks and have:
```
#elif defined(HOST_WIN32)
...
#else
...
```

File: mono/mini/mini-arm64.h [hunk 16]
```diff
@@ -15,7 +15,11 @@
 #include <mono/arch/arm64/arm64-codegen.h>
 #include <mono/mini/mini-arm64-gsharedvt.h>
 
+#ifdef HOST_WIN32
+#define MONO_ARCH_CPU_SPEC mono_arm64_desc
```
Comment (joncham) [inline mono/mini/mini-arm64.h]: Why is windows different here?

File: mono/mini/mini-arm64.h [hunk 17]
```diff
@@ -27,9 +31,15 @@
...
 #define MONO_CONTEXT_SET_LLVM_EXC_REG(ctx, exc) do { (ctx)->regs [0] = (gsize)exc; } while (0)
 
+/*
+* __builtin_frame_address () is broken on some older gcc versions in the presence of
+* frame pointer elimination, see bug #82095.
+*/
 #define MONO_INIT_CONTEXT_FROM_FUNC(ctx,func) do {	\
-		MONO_CONTEXT_SET_BP ((ctx), __builtin_frame_address (0));	\
-		MONO_CONTEXT_SET_SP ((ctx), __builtin_frame_address (0));	\
+        int tmp; \
```
Comment (joncham) [inline mono/mini/mini-arm64.h]: I would not change the non-Windows path here. On windows perhaps you can use `_AddressOfReturnAddress`?

File: mono/mini/mini-windows.c [hunk 18]
```diff
@@ -71,6 +73,81 @@ typedef struct {
 	handler handler;
 } HandlerItem;
 
+#if !defined(_MSC_VER)
+/* sigcontext surrogate */
+struct sigcontext {
+	guint64 eax;
```
Comment (joncham) [inline mono/mini/mini-windows.c]: I think this should be arch dependent

Comment (TautvydasZilys) [reply mono/mini/mini-windows.c]: I don't think it's even used, otherwise it would definitely not compile.

File: mono/utils/mono-sigcontext.h [hunk 19]
```diff
@@ -491,12 +491,20 @@ typedef struct ucontext {
...
+	#define UCONTEXT_REG_PC(ctx) (((CONTEXT*)(ctx))->Pc)
+	#define UCONTEXT_REG_SP(ctx) (((CONTEXT*)(ctx))->Sp)
+	#define UCONTEXT_REG_R0(ctx) (((CONTEXT*)(ctx))->X [ARMREG_R0])
+	#define UCONTEXT_GREGS(ctx) (&(((CONTEXT*)(ctx))->X))
 #endif
 
 #ifndef UCONTEXT_REG_SET_PC
 #define UCONTEXT_REG_SET_PC(ctx, val) do { \
-	UCONTEXT_REG_PC (ctx) = (val); \
+	UCONTEXT_REG_PC (ctx) = (guint64)(val); \
```
Comment (joncham) [inline mono/utils/mono-sigcontext.h:504]: Why does windows only need the cast?

File: mono/mini/mini-windows.c (hunk 11 above)
Comment (joncham) [reply mono/mini/mini-windows.c]: Correct, there are needed at runtime to support null reference (and other) exceptions.

Comment (joncham) [reply mono/mini/mini-windows.c]: Note, if this proves problematic or you need to debug if they are correct, you can run mono and set the `explicit-null-checks` option to have the JIT inject explicit checks rather than relying on signal handlers.

File: mono/mini/mini-windows.c (hunk 18 above)
Comment (jempatelunity) [reply mono/mini/mini-windows.c:79]: its not used, i had commited incorrectly typed syntax - but noticed it errored on other platforms but not windows

File: mono/mini/mini-windows.c (hunk 11 above)
Comment (TautvydasZilys) [reply mono/mini/mini-windows.c]: Looks like only `get_win32_restore_stack` for handling stack overflows is architecture specific. Others use plain win32 APIs so shouldn't be problematic.

File: mono/mini/mini-arm64.h (hunk 16 above)
Comment (jem-patel) [reply mono/mini/mini-arm64.h]: Naming issue. Linker error on windows. 
image

File: mono/utils/mono-sigcontext.h (hunk 19 above)
Comment (jem-patel) [reply mono/utils/mono-sigcontext.h:504]: needs the cast because if not theres issues in indirection 
image

File: mono/mini/mini-arm64.h (hunk 16 above)
Comment (TautvydasZilys) [reply mono/mini/mini-arm64.h]: Could this be because the name specified in https://github.com/Unity-Technologies/mono/pull/1695/files#diff-0a059ef6b7d69d5aa0ae445431f3bc94621d53b98ca22a033dfa51c6acd46f4b OS `arm64_desc` instead of `arm64_cpu_desc`?

Comment (jem-patel) [reply mono/mini/mini-arm64.h]: fixed in next iteration 31530c6a9ed

File: external/buildscripts/build_win_no_cygwin.pl [hunk 20]
```diff
@@ -121,22 +125,24 @@
...
+		copy("$monoroot/msvc/build/sgen/$archNameForBuild/bin/$configDirName/mono-sgen.exe", "$monoprefix/bin/.") or die ("failed copying mono-sgen.exe\n");
+		copy("$monoroot/msvc/build/sgen/$archNameForBuild/bin/$configDirName/mono-2.0-sgen.dll", "$monoprefix/bin/.") or die ("failed copying mono-2.0-sgen.dll\n");
+		copy("$monoroot/msvc/build/sgen/$archNameForBuild/bin/$configDirName/mono-2.0-sgen.pdb", "$monoprefix/bin/.") or die ("failed copying mono-2.0-sgen.pdb\n");
+		
+		# sgen as default exe
+		copy("$monoroot/msvc/build/sgen/$archNameForBuild/bin/$configDirName/mono-sgen.exe", "$monoprefix/bin/mono.exe") or die ("failed copying mono-sgen.exe to mono.exe\n");
+	}
+	
 	copy("$monoroot/msvc/build/boehm/$archNameForBuild/bin/$configDirName/MonoPosixHelper.dll", "$monoprefix/bin/.") or die ("failed copying MonoPosixHelper.dll\n");
 	copy("$monoroot/msvc/build/boehm/$archNameForBuild/bin/$configDirName/MonoPosixHelper.pdb", "$monoprefix/bin/.") or die ("failed copying MonoPosixHelper.pdb\n");
```
Comment (TautvydasZilys) [inline external/buildscripts/build_win_no_cygwin.pl]: It would be nice to get these also into the "not ARM64" check, as right now the built zip file contains just these two files in the bin directory (which is kind of pointless).

File: external/buildscripts/build_win_no_cygwin.pl [hunk 21]
```diff
@@ -199,29 +226,35 @@
...
-	copy("$monoprefix/bin/mono-sgen.exe", "$distDirArchBin/.") or die ("failed copying mono-sgen.exe\n");
-	copy("$monoprefix/bin/mono-bdwgc.exe", "$distDirArchBin/.") or die ("failed copying mono-bdwgc.exe\n");
-	copy("$monoprefix/bin/mono.exe", "$distDirArchBin/.") or die ("failed copying mono.exe\n");
+		copy("$monoprefix/bin/mono-sgen.exe", "$distDirArchBin/.") or die ("failed copying mono-sgen.exe\n");		
+		copy("$monoprefix/bin/mono-bdwgc.exe", "$distDirArchBin/.") or die ("failed copying mono-bdwgc.exe\n");
+		copy("$monoprefix/bin/mono.exe", "$distDirArchBin/.") or die ("failed copying mono.exe\n");
+	}
 
 	copy("$monoprefix/bin/MonoPosixHelper.dll", "$distDirArchBin/.") or die ("failed copying MonoPosixHelper.dll\n");
 	copy("$monoprefix/bin/MonoPosixHelper.pdb", "$distDirArchBin/.") or die ("failed copying MonoPosixHelper.pdb\n");
```
Comment (TautvydasZilys) [inline external/buildscripts/build_win_no_cygwin.pl]: And these too.

File: mono/mini/mini-arm64.h [hunk 22]
```diff
@@ -27,11 +27,26 @@
 
 #define MONO_CONTEXT_SET_LLVM_EXC_REG(ctx, exc) do { (ctx)->regs [0] = (gsize)exc; } while (0)
 
+/*
+* __builtin_frame_address () is broken on some older gcc versions in the presence of
+* frame pointer elimination, see bug #82095.
+*/
```
Comment (TautvydasZilys) [inline mono/mini/mini-arm64.h]: What's this comment from? Did you accidentally copy it from Intel code path?

File: mono/mini/mini-windows.c [hunk 23]
```diff
@@ -71,6 +75,201 @@ typedef struct {
...
+	}
+#endif
+
+	if (mono_dump_start ())
+		mono_handle_native_crash (mono_get_signame (SIGSEGV), NULL, NULL);
+
+	return EXCEPTION_CONTINUE_SEARCH;
+}
+
+#if HAVE_API_SUPPORT_WIN32_RESET_STKOFLW && !TARGET_ARM64
```
Comment (TautvydasZilys) [inline mono/mini/mini-windows.c:109]: Let's create a bug so that we don't forget that ARM64 will be missing implementation of this function.

**Review by TautvydasZilys — APPROVED** [review]
Very nice job!

File: mono/mini/mini-windows.c (hunk 23 above)
Comment (jem-patel) [reply mono/mini/mini-windows.c:109]: https://jira.unity3d.com/browse/UUM-20941

File: external/buildscripts/build_win_no_cygwin.pl [hunk 24]
```diff
@@ -121,24 +125,26 @@
...
-
-	# sgen as default exe
-	copy("$monoroot/msvc/build/sgen/$archNameForBuild/bin/$configDirName/mono-sgen.exe", "$monoprefix/bin/mono.exe") or die ("failed copying mono-sgen.exe to mono.exe\n");
-
-	copy("$monoroot/msvc/build/boehm/$archNameForBuild/bin/$configDirName/MonoPosixHelper.dll", "$monoprefix/bin/.") or die ("failed copying MonoPosixHelper.dll\n");
-	copy("$monoroot/msvc/build/boehm/$archNameForBuild/bin/$configDirName/MonoPosixHelper.pdb", "$monoprefix/bin/.") or die ("failed copying MonoPosixHelper.pdb\n");
+	
+	if($targetArch ne 'ARM64')
+	{
+		copy("$monoroot/msvc/build/sgen/$archNameForBuild/bin/$configDirName/mono-sgen.exe", "$monoprefix/bin/.") or die ("failed copying mono-sgen.exe\n");
```
Comment (joncham) [inline external/buildscripts/build_win_no_cygwin.pl:138]: Nit, are there whitespace issues here with indenting?

File: mono/mini/mini-windows.c (hunk 23 above)
Comment (joncham) [reply mono/mini/mini-windows.c:109]: Can you do a positive check for `TARGET_AMD64` rather than a negative check of `!TARGET_ARM64`?

Not blocking, but there is Win32 code path for this as well. If we are extracting these to this windows specific file, it would be nice to consolidate them all.

File: mono/mini/mini-windows.c [hunk 25]
```diff
@@ -301,7 +500,11 @@ thread_timer_expired (HANDLE thread)
 		guchar *ip;
 
 #ifdef _WIN64
-		ip = (guchar *) context.Rip;
+#ifdef TARGET_ARM64
+		ip = (guchar*)context.Pc;
+#else //TARGET_AMD64
```
Comment (joncham) [inline mono/mini/mini-windows.c]: Nit, can you do positive check here for `TARGET_AMD64` and error in the else case?

File: mono/utils/mono-threads-coop.c [hunk 26]
```diff
@@ -198,7 +198,7 @@ copy_stack_data_internal (MonoThreadInfo *info, MonoStackData *stackdata_begin,
 	state->gc_stackdata_size = stackdata_size;
 }
 
-#ifdef _MSC_VER
+#if defined(_MSC_VER) && !defined(TARGET_ARM64)
```
Comment (joncham) [inline mono/utils/mono-threads-coop.c:201]: So this block has no implementation under ARM64 and will lead to linker error (for `copy_stack_data_internal_win32_wrapper`) if used? I think that's fine, just confirming understanding.

File: mono/utils/mono-tls.c [hunk 27]
```diff
@@ -31,7 +31,7 @@
 #define MONO_THREAD_VAR_OFFSET(var,offset) __asm ("movl $" #var "@ntpoff, %0" : "=r" (offset))
 #endif
 
-#elif defined(TARGET_ARM64) && !defined(PIC)
+#elif defined(TARGET_ARM64) && !defined(PIC) && !defined(HOST_WIN32)
```
Comment (joncham) [inline mono/utils/mono-tls.c:34]: Doesn't need addressed in this PR, but I'd like to understand state of TLS for current mono platforms. There are a few arch checks later in the file, and it's unclear to me what windows on amd64/x86 is using and how ARM64 will compare to that.

File: mono/utils/winarm64.asm [hunk 28]
```diff
@@ -0,0 +1,48 @@
+	AREA	|.text|,CODE
+
+	EXPORT	|mono_context_get_current|
+
+|mono_context_get_current|	PROC
+	mov	x16, x0
```
Comment (joncham) [inline mono/utils/winarm64.asm:7]: Nit, a little bit of comments here might help

File: msvc/eglibarm64.vcxproj [hunk 29]
```diff
@@ -0,0 +1,213 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (joncham) [inline msvc/eglibarm64.vcxproj:1]: Where is this project needed? Beyond `msvc/eglib.vcxproj`?

**Comment by joncham** [comment]
Great work @jempatelunity ! I left a few comments, but I don't recall any of them were a strong blocker.

Question: does this work include using mono as executable (build tools or utilities from Editor), or just for embedding within Unity? Right now, we run the sgen variant of mono at command line AFAIK. in the ARM64 case we'd need the boehm one as I understand. But ideally, we don't use mono executable at all anymore for tools.

Testing: I think one of the best options for test coverage and stability would be to run the IL2CPP tests using this ARM64 Mono as a "comparison" run. I'm not sure of what devices/images we currently have for this on Yamato. cc @TautvydasZilys @mrvoorhe @joshpeterson

File: msvc/eglibarm64.vcxproj (hunk 29 above)
Comment (jem-patel) [reply msvc/eglibarm64.vcxproj:1]: lib created from this is used in monow.vcxproj, test_eglib.vcxproj, libtest.vcxproj and monoposixhelper.vcxproj

File: mono/utils/mono-threads-coop.c (hunk 26 above)
Comment (jem-patel) [reply mono/utils/mono-threads-coop.c:201]: yes

File: msvc/eglibarm64.vcxproj (hunk 29 above)
Comment (joncham) [reply msvc/eglibarm64.vcxproj:1]: But why is eglib.vcxproj not enough?

Comment (jem-patel) [reply msvc/eglibarm64.vcxproj:1]: oh sorry, because lib from eglib.vcxproj is x64 which is needed to run genmdesc to create arm64 files and the arm64 lib from this project is used in the projects mentioned above

File: mono/utils/winarm64.asm (hunk 28 above)
Comment (jem-patel) [reply mono/utils/winarm64.asm:7]: sure, what kind of comments are you looking for?

Comment (jempatelunity) [reply mono/utils/winarm64.asm:6]: image

Comment (jem-patel) [reply mono/utils/winarm64.asm:7]: image

**Comment by jem-patel** [comment]
@joncham This work does not include mono as executable, it is just for embedding within unity - but i guess this isnt needed as you mentioned not using mono as an executable anymore?

In regards to testing, Tom Toye will be testing a branch with various test projects to look for bugs. However there are no windows arm64 kits on bokken for katana or yamato to utilise, if we can run those tests locally let me know and I can pass them to Tom to try out?

**Comment by TautvydasZilys** [comment]
> or just for embedding within Unity? 

Specifically, this work is aiming to get it working enough for usage in the player, as that's what we're shipping in 2023.1. Editor isn't shipping until 2024.1 so it isn't a priority right this second.

> Right now, we run the sgen variant of mono at command line AFAIK. 

I think we will just switch everything to .NET if we find any mono executable usages when it's time to port the editor.

> I think one of the best options for test coverage and stability would be to run the IL2CPP tests using this ARM64 Mono as a "comparison" run. 

That would be great but my understanding is that it's not trivial to do this locally given how many suites IL2CPP has. We were aiming to run the player tests itself (runtime tests mostly) to verify it at a surface level. Unfortunately bokken team won't get these devices onto the build farm until late Q1 or even Q2 of next year so our options at automated testing are kind of limited. Given that the macOS ARM64 version is functional, we were hoping we'd just need to test enough to exercise the Windows specific paths to get us in a minimum viable testing state. Fortunately, unlike with x64, ARM64 shares calling conventions and ABI between Windows and Mac.

File: mono/utils/winarm64.asm (hunk 28 above)
Comment (TautvydasZilys) [reply mono/utils/winarm64.asm:7]: The way this was written:

1. gcc/clang macro that used inline assembly was compiled for Android with Clang;
3. The compiled object file was disassembled;
4. The resulting assembly was copy pasted into this file.

File: msvc/eglibarm64.vcxproj (hunk 29 above)
Comment (TautvydasZilys) [reply msvc/eglibarm64.vcxproj:1]: In case what Jem said isn't entirely clear, here's a diagram:

                       -----------> mono-bdwgc (ARM64) -> eglib (ARM64)
                     /
                 Mono.sln
                     \
                       -----------> genmdesc (x64) -> eglib (x64)

 The way I understand it it's because both x64 and ARM64 versions of the library are needed in the same build graph and MSBuild doesn't support that from a single project file.

Comment (joncham) [reply msvc/eglibarm64.vcxproj:1]: Thanks, makes sense!

**Review by joncham — APPROVED** [review]

**Review by TautvydasZilys — APPROVED** [review]
